src/api/api.js

```diff
@@ -36,78 +35,85 @@ api_schema.register_api(require('./func_api'));
 api_schema.compile();
 
-class APIClient {
-
-    constructor(rpc, default_options) {
-        this.rpc = rpc;
-        this.options = _.create(default_options);
-        this.RPC_BUFFERS = RPC_BUFFERS;
-
-        // define the client properties
-        this.auth = undefined;
-        this.account = undefined;
-        this.system = undefined;
-        this.tier = undefined;
-        this.node = undefined;
-        this.host = undefined;
-        this.bucket = undefined;
-        this.events = undefined;
-        this.object = undefined;
-        this.agent = undefined;
-        this.block_store = undefined;
-        this.stats = undefined;
-        this.scrubber = undefined;
-        this.debug = undefined;
-        this.redirector = undefined;
-        this.tiering_policy = undefined;
-        this.pool = undefined;
-        this.cluster_server = undefined;
-        this.cluster_internal = undefined;
-        this.server_inter_process = undefined;
-        this.hosted_agents = undefined;
-        this.frontend_notifications = undefined;
-        this.func = undefined;
-        this.func_node = undefined;
-
-        _.each(rpc.schema, api => {
-            if (!api || !api.id || api.id[0] === '_') return;
-            const name = api.id.replace(/_api$/, '');
-            if (name === 'rpc' || name === 'options') throw new Error('ILLEGAL API ID');
-            this[name] = {};
-            _.each(api.methods, (method_api, method_name) => {
-                this[name][method_name] = (params, options) => {
-                    options = _.create(this.options, options);
-                    return rpc._request(api, method_api, params, options);
-                };
-            });
-        });
-    }
+const shared_client_proto = {
+    RPC_BUFFERS,
 
-    /**
-     * extend the rpc client prototype with convinient methods
-     */
-    create_auth_token(params) {
-        return this.auth.create_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+    create_auth_token: async function(params) {
+        const res = this.auth.create_auth(params);
+        this.options.auth_token = res.token;
+        return res;
+    },
 
-    create_access_key_auth(params) {
-        return this.auth.create_access_key_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+    create_access_key_auth: async function(params) {
+        const res = this.auth.create_access_key_auth(params);
+        this.options.auth_token = res.token;
+        return res;
+    },
 
-    async create_k8s_auth(params) {
+    create_k8s_auth: async function(params) {
         const res = await this.auth.create_k8s_auth(params);
         this.options.auth_token = res.token;
         return res;
+    },
+
+    _invoke_api(api, method_api, params, options) {
+        options = Object.assign(
```

Comment:
Isn't `Object.create(this.options, options)` doing that?

Reply:
No, `Object.create`second parameter is an object properties descriptor, not a plain object. What we want is to copy the options from `options` into the newly created object, this is exactly what Object.assign was designed to do.

Reply:
I probably mixed up with the api of https://lodash.com/docs/4.17.15#create which was used in this code before your changes to create the combined options `options = _.create(this.options, options);`